tensorflow_recommenders_addons/dynamic_embedding/python/ops/shadow_embedding_ops.py

```diff
@@ -0,0 +1,213 @@
+    ids_name = self._name + '-ids'
+    if ids is None:
+      self.ids = resource_variable_ops.ResourceVariable(
+          (),
+          trainable=False,
+          collections=collections,
+          name=ids_name,
+          dtype=self.params.key_dtype,
+          shape=tensor_shape.TensorShape(None))
+    else:
+      if not isinstance(ids, resource_variable_ops.ResourceVariable):
+        raise TypeError('If ids is set, it needs to be a ResourceVariable')
+      self.ids = ids
+    if kwargs.get('create_as_slot', False):
+      kwargs.pop('create_as_slot')
+    else:
+      self._track_trackable(self.ids, ids_name, overwrite=False)
+    model_mode = kwargs.get('model_mode', None)
+    if model_mode:
+      kwargs.pop('model_mode')
+    else:
+      model_mode = de.ModelMode.CURRENT_SETTING
+    initial_value = array_ops.zeros(shape=(0, self.params.dim),
+                                    dtype=self.params.value_dtype)
+    distribute_strategy = kwargs.get('distribute_strategy', None)
+    super(ShadowVariable,
+          self).__init__(params,
+                         self.ids,
+                         max_norm=max_norm,
+                         initial_value=initial_value,
+                         dtype=self.params.value_dtype,
+                         trainable=trainable,
+                         collections=collections,
+                         model_mode=model_mode,
+                         distribute_strategy=distribute_strategy,
+                         name=name)
+    exists = kwargs.get('exists', None)
+    if exists is None:
+      self.exists = resource_variable_ops.ResourceVariable(
+          (),
+          trainable=False,
+          collections=collections,
+          name=ids_name,
+          dtype=dtypes.bool,
+          shape=tensor_shape.TensorShape(None))
+    else:
+      self.exists = exists
+
+  def prefetch_values(self, update=False):
+    if self.params.bp_v2:
+      r, exists = self.params.lookup(self.ids, return_exists=True)
+      self.exists.assign(exists)
+      self.prefetch_values_op = self.transform(r)
+    else:
+      self.prefetch_values_op = self.transform(self.params.lookup(self.ids))
+    return self.prefetch_values_op
+
+  def _reset_ids(self, ids):
+    if not context.executing_eagerly():
+      collected_ops = []
+    with ops.control_dependencies([self.ids.assign(ids, use_locking=True)]):
+      prefetch_op = self.prefetch_values(update=True)
+      if not context.executing_eagerly():
+        collected_ops.append(prefetch_op)
+    for s in self._tracked_slots:
+      partial_reset_op = s._reset_ids(ids)
+      if not context.executing_eagerly():
+        collected_ops.append(partial_reset_op)
+    if not context.executing_eagerly():
+      return control_flow_ops.group(collected_ops)
+
+  def _gather_saveables_for_checkpoint(self):
+    self._reset_ids(array_ops.zeros((0,), dtype=self.params.key_dtype))
+    return {trackable.VARIABLE_VALUE_KEY: self}
+
+
+def embedding_lookup(
```

Comment:
So this is where you want to reuse the code that is xx_unique and xx_sparse? Is it possible to write these two functions as a decorators of embedding_lookup?

Reply:
Yes, in best case it should apply no extra changes to user code.

Reply:
> So this is where you want to reuse the code that is xx_unique and xx_sparse? Is it possible to write these two functions as a decorators of embedding_lookup?

Already support to add embedding_lookup implicitly inside tf.function